agent/command.go

```diff
@@ -146,15 +146,29 @@ func (a *Agent) runCommandOrFunc(ctx context.Context, tc *taskContext, commandIn
 // single sub-command within a function.
 func (a *Agent) runCommand(ctx context.Context, tc *taskContext, logger client.LoggerProducer, commandInfo model.PluginCommandConf,
 	cmd command.Command, displayName string, options runCommandsOptions) error {
-
+	prevExp := map[string]string{}
 	for key, val := range commandInfo.Vars {
+		prevVal := tc.taskConfig.Expansions.Get(key)
+		prevExp[key] = prevVal
+
 		var newVal string
 		newVal, err := tc.taskConfig.Expansions.ExpandString(val)
 		if err != nil {
 			return errors.Wrapf(err, "expanding '%s'", val)
 		}
 		tc.taskConfig.Expansions.Put(key, newVal)
 	}
+	defer func() {
+		// This defer ensures that the function vars do not persist in the expansions after the function is over.
+		for key, functionValue := range commandInfo.Vars {
+			currentValue := tc.taskConfig.Expansions.Get(key)
+			if currentValue != functionValue {
+				// if an expansion has been updated to something that's not a function var, persist it.
```

Comment:
nit: capitalization

Reply:
Rather than describe this in terms of function vars, I would just phrase this as something like "If the expansions.update command updates the expansion value, persist it."